src/transforms/babel.js

```diff
@@ -26,19 +21,81 @@ module.exports = async function(asset) {
   }
 };
 
-async function shouldTransform(asset) {
+async function getConfig(asset) {
+  // If asset is marked as an ES6 modules, this is a second pass after dependencies are extracted.
+  // Just compile modules to CommonJS.
   if (asset.isES6Module) {
-    return true;
+    return {
+      plugins: [require('babel-plugin-transform-es2015-modules-commonjs')]
+    };
+  }
+
+  if (asset.babelConfig) {
+    return asset.babelConfig;
+  }
+
+  // Attempt to generate a config
+  let config = await getEnvConfig(asset);
+
+  // If this is the app module, and there is a .babelrc, merge it into the config.
+  // Don't load .babelrc for node_modules. https://github.com/parcel-bundler/parcel/issues/13.
+  if (!asset.name.includes('/node_modules/')) {
+    let babelrc =
+      (asset.package && asset.package.babel) ||
+      (await asset.getConfig(['.babelrc', '.babelrc.js']));
+    if (babelrc) {
+      config = Object.assign({}, babelrc, {
+        plugins: (babelrc.plugins || []).concat(config ? config.plugins : [])
+      });
+    }
+  }
+
+  return config;
+}
+
+module.exports.getConfig = getConfig;
+
+async function getEnvConfig(asset) {
+  // Load supported source engines (node and browser versions) for source module and target app,
+  // and generate a babel-preset-env config for each.
+  let sourceEngines = await getTargetEngines(asset);
+  let targetEngines = await getTargetEngines(asset, asset.options.mainFile);
+  let sourceEnv = await getEnvPlugins(sourceEngines);
+  let targetEnv = await getEnvPlugins(targetEngines);
+
+  if (!sourceEnv || !targetEnv) {
+    return;
+  }
+
+  // Do a diff of the returned plugins. We only need to process the remaining plugins to get to the app target.
+  // If this is the app module, the source and target will be the same, so just compile everything.
+  if (asset.name.includes('/node_modules/')) {
```

Comment:
Need a nicer way to detect files that are in the app rather than a module

Reply:
there might not be any, we've discussed this recently at babel and it seems there is no better solution to see if a file is node_module or not than to look at its path, and we went with this approach for this kind of detection

Reply:
In CRA we're currently just processing anything that's *not* in application `src` folder with `-env`. Is that wrong?

Reply:
That's fine if you know something about the file structure (that there is a src folder), but Parcel needs to work for all types of projects.

Reply:
Thanks for explaining!